### PR TITLE
Deprecate reload_dataloaders_every_n_epochs flag on Trainer and set it as a property for DataHooks 

### LINKED ISSUE #8738: Move reload_dataloaders_every_n_epochs to the DataHooks class
### Motivation

We are auditing the Lightning components and APIs to assess opportunities for improvements:
- https://docs.google.com/document/d/1xHU7-iQSpp9KJTjI3As2EM0mfNHHr37WZYpDpwLkivA/edit#
- https://github.com/PyTorchLightning/pytorch-lightning/issues/7740

`reload_dataloaders_every_n_epochs ` is an argument to the Trainer constructor. However, this could be a property of the DataHooks class, instead of the trainer, as whether to initiate reloading the dataloading every n epochs should be determined by the actor providing the dataloaders (e.g. the LightningModule or LightningDataModule). 

This is very similar to #8733 and how automatic/manual optimization is a property of the LightningModule. That property also started out as a trainer argument before being migrated to the lightning module. Since this pattern keeps occurring, we should separately understand why it's so appealing to add things to the trainer constructor instead of a more specific component. 

Moreover, this one setting controls dataloader behavior for both train & val dataloaders. Do we need more granular control? Do we need two properties, one for training and one for validation? This could make sense as we could have very different epoch counts with features like `val_check_interval`, where the training epoch count != val epoch count. The property for validation would only apply during `trainer.fit` as `trainer.validate` only makes a single pass through the data. 

However, the documentation for the `test_dataloader`: https://github.com/PyTorchLightning/pytorch-lightning/blob/963c26764682fa4cf64c93c5a7572ae0040e9c32/pytorch_lightning/core/hooks.py#L535-L537
Is this a copy/paste issue? 

### Pitch

- Add a property to the DataHooks class for this in v1.5
- Deprecate the Trainer argument for this in v1.5
- Remove the Trainer argument in v1.7

Benefits:
- Simplify the Trainer constructor (one fewer argument)
- Keep the data loader management in one place instead of two (at the DataHooks level)

### Alternatives
Keep as is?

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

<sub>

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, finetuning and solving problems with deep learning

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks and more for research and production with PyTorch Lightning and PyTorch

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

</sub>

### PR DESCRIPTION
**Pull request by ninginthecloud**
## What does this PR do?

Fixes #8738
This PR has completed the following:
1. add `DataHooks` with `reload_dataloaders_every_n_epochs` property
2. set `reload_dataloaders_every_n_epochs` in `trainer` as optional
3. update `on_trainer_init` function in `data_connector` class
4. add `_should_reload_val_dl_epoch()` and `_should_reload_train_dl_epoch()` in order to introduce two new properties
`reload_train_dataloader_every_n_epochs` and `reload_val_dataloader_every_n_epochs`

## Before submitting

- [X] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [X] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [X] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [X] Did you make sure to **update the documentation** with your changes? (if necessary)
- [X] Did you write any **new necessary tests**? (not for typos and docs)
- [X] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [X] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/core/hooks.py [hunk 1]
```diff
@@ -378,9 +378,25 @@ def __init__(self) -> None:
             prepare_data_per_node:
                 If True, each LOCAL_RANK=0 will call prepare data.
                 Otherwise only NODE_RANK=0, LOCAL_RANK=0 will prepare data.
+            reload_dataloaders_every_n_epochs:
+                Set to a non-negative integer to reload dataloaders every n epochs.
+                Default: 0
+            reload_train_dataloader_every_n_epochs:
+                Set to a non-negative integer to reload train_dataloader every n epochs.
```
Comment (tchaton) [inline pytorch_lightning/core/hooks.py:385]:
```suggestion
                Set to a non-negative integer to reload train dataloader every n epochs.
```

File: pytorch_lightning/core/hooks.py [hunk 2]
```diff
@@ -378,9 +378,25 @@ def __init__(self) -> None:
...
+                Default: 0
+            reload_val_dataloader_every_n_epochs:
+                Set to a non-negative integer to reload val_dataloader every n epochs.
```
Comment (tchaton) [inline pytorch_lightning/core/hooks.py:388]:
```suggestion
                Set to a non-negative integer to reload val dataloaders every n epochs.
```

File: pytorch_lightning/loops/dataloader/evaluation_loop.py [hunk 3]
```diff
@@ -249,3 +250,27 @@ def on_evaluation_epoch_end(self) -> None:
     def teardown(self) -> None:
         self._results.cpu()
         self.epoch_loop.teardown()
+
+    def _should_reload_val_dl_epoch(self) -> bool:
+        """Check if val dataloader should be reloaded in the current epoch."""
+        current_epoch = self.epoch_loop.batch_progress.current.completed
```
Comment (tchaton) [inline pytorch_lightning/loops/dataloader/evaluation_loop.py:256]:
```suggestion
        current_epoch = self.trainer.current_epoch
```

Comment (tchaton) [reply pytorch_lightning/loops/dataloader/evaluation_loop.py:256]: This was actually self.epoch_progress.current.completed from the fit_loop.

Comment (ninginthecloud) [reply pytorch_lightning/loops/dataloader/evaluation_loop.py:256]: Yes. This is one of my question.  I noticed that original implementation is to initialize current_epoch with self.epoch_progress.current.completed from the fit_loop, and since now, we want to give users more granular control, by introducing `reload_train_dataloader_every_n_epochs` and `reload_val_dataloader_every_n_epochs `.Therefore, considering this is evaluation_loop, does it still make sense to get current epoch from fit_loop?

**Comment by tchaton** [comment]
Dear @ninginthecloud,

Any updates on this one ?

Best,
T.C

**Comment by ninginthecloud** [comment]
Hi, @tchaton, sorry about replying late. This PR will impact the interface of current trainer. I will push this deprecation for V1.6.

**Comment by carmocca** [comment]
Closing this PR as it has become stale. We would agree to merge this if somebody wants to take over the authorship. Thank you for your time!
